### PR TITLE
Separate compiler passes more explicitly.

### PR DESCRIPTION
**Pull request by bchetioui**
To be able to do that, PackageHeads also now carry source information
that can be queried. The compiler can now be called using the "test"
subcommand to test compilation up to a specific compiler pass, and
for a specific backend (once codegen will be implemented).

The commit also cleans up the legacy WithSrc-related utils, which are no
longer needed. Various additional cleanups are forthcoming, but not
included here, so as to keep the resulting PR readable.

### CODE REVIEW

File: src/magnolia.hs [hunk 1]
```diff
@@ -18,9 +17,37 @@ import PPrint
...
+runTest config filePath = case _testPass config of
+  DepAnalPass -> runAndLogErrs $ depAnalPass filePath
+  ParsePass -> runAndLogErrs $ depAnalPass filePath >>= parsePass
+  CheckPass -> runAndLogErrs $ depAnalPass filePath >>= parsePass >>= checkPass
+  CodegenPass -> error "codegen not yet implemented"
+
+-- === debugging utils ===
+
+logErrs :: S.Set Err -> IO ()
+logErrs errs = pprintList (L.sort (toList errs))
```
Comment (mariuslarnoy) [inline src/magnolia.hs:45]: Wouldn't we lose information about when the error occurred if the list gets sorted before it is printed?

**Review by mariuslarnoy — COMMENTED** [review]
Nice cleanup, and better separation of the passes makes the code easier to navigate and read in general. LGTM!

File: src/magnolia.hs (hunk 1 above)
Comment (bchetioui) [reply src/magnolia.hs:45]: Well, `errs` is a set to begin with – which makes no promises about conserving ordering of inputs (`OSet` would do that, but it didn't seem warranted), so this at least doesn't apply here.

From a more general standpoint, I don't think that there is much of a point from the point of view of the Magnolia developer to know in what order the errors occurred (?).

Another issue with not sorting is that topological sorts of our list of packages/modules are not unique, which means that we are not guaranteed to get a stable output if we do not sort the errors before printing them out. This is a problem, at the very least when it comes to our testing.

What do you think?

**Comment by bchetioui** [comment]
Thanks for your review. I will now merge.

File: src/magnolia.hs (hunk 1 above)
Comment (mariuslarnoy) [reply src/magnolia.hs:45]: Alright that clears it up for me, I was just curious about the reason for sorting in the first place. On second thought, I guess we don't really get anything useful information out of knowing the order. Thanks!

**Pull request merged**